File: Assets/Scripts/Core/GameFlow/PlayerState.cs
namespace Parkiller.Core
{
    public class PlayerState
    {
        public readonly PieceColor Color;
        public readonly Piece[] Pieces = new Piece[4];

        public PlayerState(PieceColor color)
        {
            Color = color;
            for (int i = 0; i < Pieces.Length; i++)
                Pieces[i] = new Piece(color, i);
        }

        public bool HasWon
        {
            get
            {
                foreach (var piece in Pieces)
                    if (piece.State != PieceState.Finished)
                        return false;
                return true;
            }
        }
    }
}

File: src/core/gameFlow/playerState.ts
import type { PieceColor } from '../pieceColor'
import { createPiece, type Piece } from '../pieces/piece'

export interface PlayerState {
  color: PieceColor
  pieces: Piece[]
}

export function createPlayerState(color: PieceColor): PlayerState {
  return {
    color,
    pieces: [0, 1, 2, 3].map((i) => createPiece(color, i)),
  }
}

export function hasWon(player: PlayerState): boolean {
  return player.pieces.every((p) => p.state === 'Finished')
}
